### PR TITLE
Adding the grain exercise.

### PR DESCRIPTION
**Pull request by Kineolyan**
Opening the PR to mark the issue as in-progress.

Closes: #1744.

---

Reviewer Resources:

[Track Policies](https://github.com/exercism/java/blob/master/POLICIES.md#event-checklist)

### CODE REVIEW

File: exercises/grains/src/test/java/GrainsTest.java [hunk 1]
```diff
@@ -0,0 +1,88 @@
...
+import java.math.BigInteger;
+
+import static org.junit.Assert.assertEquals;
+import static org.junit.Assert.fail;
+
+public class GrainsTest {
+
+    @Test
+    public void countAtCell1() {
+        final BigInteger result = Grains.compute(1);
```
Comment (lemoncurry) [inline exercises/grains/src/test/java/GrainsTest.java]: Please avoid using `final` as specified in the [policies](https://github.com/exercism/java/blob/master/POLICIES.md#starter-implementations).

File: exercises/grains/build.gradle [hunk 2]
```diff
@@ -0,0 +1,19 @@
+apply plugin: "java"
+apply plugin: "eclipse"
+apply plugin: "idea"
+
+repositories {
+  mavenCentral()
+}
+
+dependencies {
+  testCompile group: 'org.hamcrest', name: 'hamcrest-library', version: '1.3'
```
Comment (lemoncurry) [inline exercises/grains/build.gradle]: Please remove this dependency.

**Comment by Kineolyan** [comment]
Thanks for the review. I have addressed all your comments.

File: config.json [hunk 3]
```diff
@@ -1,4 +1,4 @@
-{
+/{
```
Comment (lemoncurry) [inline config.json]: Please remove "/"

File: config.json [hunk 4]
```diff
@@ -1441,6 +1441,18 @@
       "difficulty": 0,
       "topics": null,
       "deprecated": true
+    },
+    {
+      "slug": "grains",
+      "uuid": "5ee66f39-5e37-4907-a6d9-f55d38324c6c",
+      "core": false,
+      "unlocked_by": "hello-world",
+      "difficulty": 1,
```
Comment (lemoncurry) [inline config.json]: According to the [policies](https://github.com/exercism/java/blob/master/POLICIES.md#reference-tutorial-in-the-first-exercises), you need to reference the hello world tutorial.

File: exercises/grains/.meta/src/reference/java/Grains.java [hunk 5]
```diff
@@ -0,0 +1,25 @@
...
+        }
+    }
+
+    static BigInteger computeTotalNumberOfGrainsOnBoard() {
+        return IntStream.rangeClosed(1, 64)
+                .mapToObj(Grains::computeNumberOfGrainsOnSquare)
+                .reduce(
+                        BigInteger.valueOf(0),
+                        BigInteger::add);
+        // Note that the following is also correct, using the sum of the elements of a geometric series
```
Comment (lemoncurry) [inline exercises/grains/.meta/src/reference/java/Grains.java]: Please remove comment. This is just the reference implementation, but it is of course not the only way a student may implement it.

File: exercises/grains/src/test/java/GrainsTest.java [hunk 6]
```diff
@@ -0,0 +1,88 @@
...
+        assertEquals(new BigInteger("1"), result);
+    }
+
```
[the remark this answers is not in the document]
Comment (lemoncurry) [reply exercises/grains/src/test/java/GrainsTest.java]: The annotaion needs to come before the `@Test`annotation.

**Comment by Kineolyan** [comment]
I reviewed the evaluation of the exercise, given that exercises with difficulty 1 are far more simple. This looks similar to darts, which is currently level 2.

File: config.json [hunk 7]
```diff
@@ -1441,6 +1441,18 @@
...
+      "uuid": "5ee66f39-5e37-4907-a6d9-f55d38324c6c",
+      "core": false,
+      "unlocked_by": "two-fer",
+      "difficulty": 2,
+      "topics": [
+        "loops",
+        "math"
+      ],
+      "deprecated": false
     }
```
Comment (lemoncurry) [inline config.json]: What I wanted to say was is that in this file, we group exercises by difficulty level.
You put this new exercise at the end of the file, put it should be placed with the exercises of level 2, for example on line 97.

Comment (Kineolyan) [reply config.json]: Ah, got it. I was also wondering if what was the ordering of this file, as it was not by alphabetical order.

File: exercises/grains/src/test/java/GrainsTest.java [hunk 8]
```diff
@@ -0,0 +1,99 @@
+import org.junit.Ignore;
+import org.junit.Test;
+
+import java.math.BigInteger;
+
+import static org.junit.Assert.assertEquals;
+import static org.junit.Assert.fail;
+
+public class GrainsTest {
+
```
Comment (lemoncurry) [inline exercises/grains/src/test/java/GrainsTest.java]: I would prefer a non-static implementation. See for example the pangram exercise.
This is policy, sorry I missed it in my previous comments.

Comment (Kineolyan) [reply exercises/grains/src/test/java/GrainsTest.java]: No problem. I will change this to a standard class.

**Comment by Kineolyan** [comment]
@lemoncurry maybe just having to use `BigInteger` was weird, as it was the first time for me in all my Java experience. Otherwise, nothing. The exercise is well explained and easy for new comers.
I would like to thank you for your patience, as I had missed many rules of the policies :smile:

File: exercises/grains/src/test/java/GrainsTest.java [hunk 9]
```diff
@@ -0,0 +1,108 @@
...
+    public void countAtSquare64() {
+        BigInteger result = grains.computeNumberOfGrainsOnSquare(64);
+        assertEquals(new BigInteger("9223372036854775808"), result);
+    }
+
+    @Ignore("Remove to run test")
+    @Test
+    public void errorOnNullBoardSize() {
+        try {
+            grains.computeNumberOfGrainsOnSquare(0);
```
Comment (FridaTveit) [inline exercises/grains/src/test/java/GrainsTest.java]: Please use [ExpectedException](https://junit.org/junit4/javadoc/4.12/org/junit/rules/ExpectedException.html). You can look at the [phone-number tests](https://github.com/exercism/java/blob/a5c8c2144a25b13feeb884cf6e1eeb71097b09ac/exercises/phone-number/src/test/java/PhoneNumberTest.java) for an example of how it's used 🙂

Comment (Kineolyan) [reply exercises/grains/src/test/java/GrainsTest.java]: I read about it in the notes and somehow thought it was to test exception generation, as I do in my personal project. I should have look for examples in the code.
I am updating the code right now.

Comment (Kineolyan) [reply exercises/grains/src/test/java/GrainsTest.java]: Yes, I read about it and somehow confused it with the custom exception.
By the way, this was removed in JUnit5, in favour of a new method `Assertions.assertThrows`. But that's a topic for another time.

Comment (FridaTveit) [reply exercises/grains/src/test/java/GrainsTest.java]: Good to know! It's probably worth us updating the exercises to use that at some point then 🙂

**Pull request merged**